### PR TITLE
fix(core): fix reflectConstructorParams mutating constructor metadata

### PR DESCRIPTION
**Pull request by nathan-knight**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Other... Please describe:

## What is the current behavior?

Currently `reflectConstructorParams` mutates the original constructor parameter type metadata when applying the parameter decorator information.

## What is the new behavior?
It uses a shallow copy of the metadata so it doesn't mutate the original.

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

## Other information

### CODE REVIEW

File: packages/core/test/injector/injector.spec.ts [hunk 1]
```diff
@@ -838,6 +839,20 @@ describe('Injector', () => {
       expect(dependencies).to.deep.eq([FixtureDep1, FixtureDep2]);
       expect(optionalDependenciesIds).to.deep.eq([1]);
     });
+
+    it('should not mutate the constructor metadata', async () => {
+      class FixtureDep1 {}
+      const injectionToken = 'test_token';
+      @Injectable()
+      class FixtureClass {
+        constructor(@Inject(injectionToken) private dep1: FixtureDep1) {}
```
Comment (micalevisk) [inline packages/core/test/injector/injector.spec.ts:848]: do we really need to use `@Inject()` here? :thinking: 

If yes, can we have this instead:

```
constructor(@Inject(FixtureDep1) private dep1: FixtureDep1)
```

Comment (nathan-knight) [reply packages/core/test/injector/injector.spec.ts:848]: Yes, it needs to be something other than the class itself or the test would pass with a false negative were this to regress

Comment (micalevisk) [reply packages/core/test/injector/injector.spec.ts:848]: could you please add a comment right above the `const injectionToken` to clarify that?

File: packages/core/test/injector/injector.spec.ts [hunk 2]
```diff
@@ -838,6 +839,20 @@ describe('Injector', () => {
...
+      }
+      const wrapper = new InstanceWrapper({ metatype: FixtureClass });
+      const [dependencies] = injector.getClassDependencies(wrapper);
+      expect(dependencies).to.deep.eq([injectionToken]);
+      const paramtypes = Reflect.getMetadata(PARAMTYPES_METADATA, FixtureClass);
+      expect(paramtypes).to.deep.eq([FixtureDep1]);
```
Comment (micalevisk) [inline packages/core/test/injector/injector.spec.ts:854]:
```suggestion

      const paramtypes = Reflect.getMetadata(PARAMTYPES_METADATA, FixtureClass);
      expect(paramtypes).to.deep.eq([FixtureDep1]);
```

File: packages/core/test/injector/injector.spec.ts (hunk 1 above)
Comment (nathan-knight) [reply packages/core/test/injector/injector.spec.ts:850]: Added a comment

File: packages/core/test/injector/injector.spec.ts (hunk 2 above)
Comment (nathan-knight) [reply packages/core/test/injector/injector.spec.ts:858]: I cleaned up the spacing here and in a couple other spots

**Review by micalevisk — APPROVED** [review]

File: packages/core/test/injector/injector.spec.ts [hunk 3]
```diff
@@ -10,6 +10,7 @@ import { NestContainer } from '../../injector/container';
 import { Injector, PropertyDependency } from '../../injector/injector';
 import { InstanceWrapper } from '../../injector/instance-wrapper';
 import { Module } from '../../injector/module';
+import { PARAMTYPES_METADATA } from '@nestjs/common/constants';
 chai.use(chaiAsPromised);
```
[the remark this answers is not in the document]
Comment (nathan-knight) [reply packages/core/test/injector/injector.spec.ts:15]: Hey @kamilmysliwiec, were there any other issues in the PR that need to be addressed?

**Comment by nathan-knight** [comment]
@kamilmysliwiec are there any other changes I should make so we can get this in?

**Comment by kamilmysliwiec** [comment]
lgtm
